AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,276 @@
+
+    if (((AVPlayerLayer *)_playerNode.layer).player) {
+      [((AVPlayerLayer *)_playerNode.layer).player replaceCurrentItemWithPlayerItem:_currentItem];
+    } else {
+      ((AVPlayerLayer *)_playerNode.layer).player = [[AVPlayer alloc] initWithPlayerItem:_currentItem];
+    }
+
+  }
+  if (_shouldAutoPlay) {
+    [self play];
+  }
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  
+  {
+    ASDN::MutexLocker l(_lock);
+    ((AVPlayerLayer *)_playerNode.layer).player = nil;
+    _shouldBePlaying = NO;
+  }
+}
+
+- (void)setPlayButton:(ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  _playButton = playButton;
+  
+  [self addSubnode:playButton];
+  
+  [_playButton addTarget:self action:@selector(play) forControlEvents:ASControlNodeEventTouchUpInside];
+}
+
+- (ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  return _playButton;
+}
+
+- (void)setAsset:(AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if (ASObjectIsEqual(((AVURLAsset *)asset).URL, ((AVURLAsset *)_asset).URL)) {
+    return;
+  }
+  
+  _asset = asset;
+  
+  if (self.interfaceState & ASInterfaceStateFetchData) {
+    [self fetchData];
+  }
+}
+
+- (AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  return _asset;
+}
+
+- (void)setGravity:(ASVideoGravity)gravity {
+  ASDN::MutexLocker l(_lock);
+  
+  switch (gravity) {
+    case ASVideoGravityResize:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResize;
+      break;
+    case ASVideoGravityResizeAspect:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspect;
+      break;
+    case ASVideoGravityResizeAspectFill:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspectFill;
+      break;
+    default:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspect;
+      break;
+  }
+}
+
+- (ASVideoGravity)gravity;
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if ([((AVPlayerLayer *)_playerNode.layer).contentsGravity isEqualToString:AVLayerVideoGravityResize]) {
+    return ASVideoGravityResize;
+  }
+  if ([((AVPlayerLayer *)_playerNode.layer).contentsGravity isEqualToString:AVLayerVideoGravityResizeAspectFill]) {
+    return ASVideoGravityResizeAspectFill;
+  }
+  
+  return ASVideoGravityResizeAspect;
+}
+
+- (void)play;
+{
+  ASDN::MutexLocker l(_lock);
+  
+  [[((AVPlayerLayer *)_playerNode.layer) player] play];
+  _shouldBePlaying = YES;
+  _playButton.alpha = 0.0;
+  if ([self ready] && ![self.subnodes containsObject:_spinner]) {
+    _spinner = [[ASDisplayNode alloc] initWithViewBlock:^UIView *{
+      UIActivityIndicatorView *spinnnerView = [[UIActivityIndicatorView alloc] initWithFrame:_playButton.frame];
+      spinnnerView.color = [UIColor whiteColor];
+      [spinnnerView startAnimating];
```

Comment:
It's fine to wrap this as a UIActivityIndicatorView, but technically we don't want to always call startAnimating just when the view loads.  Make sure to check the usage of _spinner very carefully to ensure that it is impossible for it to be left animating when it is offscreen, as this is one of a few classes that will continue to spend CPU resources (low level but measurable) even when outside of a hierarchy.

Also it is not quite right to use initWithFrame: in the view block, as theoretically if we were out of a node hierarchy when this is called the addSubnode: will not trigger the view block to be called, and it may end up getting called before a layout pass has set the frame on _playButton.  You can instead just do _spinner.frame = in your -layout method.